Fix the inconsistency of `polygamma` on Tensor and Variable, for issue #4466

sighingnow opened the pull request:
In issue #4466, @fritzo found that the `torch.polygamma` works inconsistently on `Tensor` and `Variable`. When call the module methods on variables, the first two arguments need to be swapped if the `self`  argument comes second.

This patch call `IMPLEMENT_STATELESS_ADDXX` instead of `IMPLEMENT_STATELESS` for the swap action before dispatch, should fix the issue #4466.

sighingnow committed:
Fix the inconsistency of `polygamma` on Tensor and Variable.

Signed-off-by: HE, Tao <sighingnow@gmail.com>

fritzo commented:
Thanks for fixing this! Do you think it's worth adding a one-line regression test?
```py
torch.polygamma(0, torch.autograd.Variable(x, requires_grad=True))
```

sighingnow commented:
Where should I put the test case? In `test/test_torch` or in `test/test_autograd` ?

sighingnow committed:
Regression test for #4466, polygamma works on variables.

Signed-off-by: HE, Tao <sighingnow@gmail.com>

sighingnow commented:
Regression test done.

apaszke commented on torch/csrc/Module.cpp:
```diff
@@ -369,6 +368,10 @@ IMPLEMENT_STATELESS_ADDXX(baddbmm)
 
+// For torch.polygamma(n, x), the `self` argument comes second, the
+// first two arguments needs to be swapped before dispatch.
+IMPLEMENT_STATELESS_ADDXX(polygamma)
```
This macro (`IMPLEMENT_STATELESS_ADDXX`) mentions that it's used for deprecated `add**` functions, which is no longer true. Can you mention that, and possibly rename it (e.g. to `IMPLEMENT_STATELESS_SWAP`) to make it clear that it shouldn't go away?

sighingnow replied:
I have add a macro `IMPLEMENT_STATELESS_SWAP` and clarify the it's usage in comment.

Now `IMPLEMENT_STATELESS_ADDXX` is an alias of `IMPLEMENT_STATELESS_SWAP` for deprecated `add**` methods.

sighingnow committed:
Add macro IMPLEMENT_STATELESS_SWAP to dispatch stateless methods on Variables correctly.

When call stateless methods with more than one arguments and the `self` comes second,
the `self` argument needs to be swapped to the first position before dispatching.

The macro `IMPLEMENT_STATELESS_ADDXX` is still reserved for deprecated `add**`
methods.

Signed-off-by: HE, Tao <sighingnow@gmail.com>

sighingnow commented:
ping @apaszke

apaszke commented:
@pytorchbot test this please

sighingnow commented:
The windows CI seems be blocked. The "test and push" phase of [build # 203](https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-trigger/203/) has consumed almost **6 hours**.

![image](https://user-images.githubusercontent.com/7144772/34726436-f685750a-f58e-11e7-9a1e-69b239e3b13d.png)

soumith commented:
thanks @sighingnow !
